### ISSUE TITLE
Not very resonable behavior of the `modify_count` in `show stats_meta`.

### ISSUE THREAD

**Issue by winoros** [pingcap/tidb#23453]
## Feature Request

**Describe the feature you'd like:**

Currently, when we analyze a big table(e.g. There are billions rows in it). The analyze will last a long time.
It cause two problems:
1. We will clean the modify count after the analyze is finished. But since the analyze lasts a long time, the modify count after it's triggered and before it's finished will lose. This would lead to wrong row count estimation.
1. Each index task is asynchronous. But each task will clean the modify count once it's finished. So this will also make the modify_count wrongly updated, which would lead to wrong row count estimation.

**Describe alternatives you've considered:**

A clean design to let the modify count be updated more correctly to avoid such scenarios.

**Teachability, Documentation, Adoption, Migration Strategy:**

**Comment by winoros** [comment]
Whatever the #23472 's test satisfies our needs or not. The `1.` will always exist. If we analyze on a large table whose data is billions. There's always a gap between the trigger time and the finish time.

**Comment by coocood** [comment]
We can do this:

When a analyze task begins to run, it read the current modified count, then after the analyze task finish it update modfied count by subtract the modified count with the value it previously read.

**Comment by dbsid** [comment]
#21207 is trying to improve the selectivity but still rely on the modify_count, the fix does not help on this issue.

For the out-of-range date range selectivity, maybe a better solution is to Use a linear decay strategy as oracle adopts. This approach does not rely on the modify_count, by rewriting the outOfRangeIntervalSelectivity using this method, we don't need to change the current statistics analyze logic.

See the the explanation from the below paper.
```
In our example, the known range is (0,199), which Oracle extends to add (199,398) and (-199,0).
Oracle then assumes that the cardinality estimates should decrease linearly from 5,000 to 0 across the
extended ranges. So, for our predicate “columnX = 250” we can then read off the graph (or do the
arithmetic) that the estimated cardinality is 3,719. (Informally, and approximately, 250 is roughly
quarter of the way from 199 to 398, so the estimated cardinality should drop roughly by a quarter of
the original 5,000.) 
```

![image](https://user-images.githubusercontent.com/2316425/113893106-4a0dc780-97f9-11eb-817c-eb4eb145b251.png)

https://www.doag.org/formes/pubfiles/6315103/2014-DB-Jonathan_Lewis-Calculating_Basic_Selectivity-Manuskript.pdf

**Comment by dbsid** [comment]
The current out of range judge is not very strict, either lowVal or highValue is out of range, even the predicate range is part of overlapping with the histogram.

This means every time the stats analyze completes and the modify_count is reset to 0, there is a risk the plan will be changed since the selectivity for out-of-range predicate will drop to zero, no matter how big or small the overlapping range is.

```
		if (c.outOfRange(lowVal) && !lowVal.IsNull()) || c.outOfRange(highVal) {
			cnt += outOfRangeIntervalSelectivity(lowVal, highVal, &c.Histogram, modifyCount)

		}
```

**Comment by eurekaka** [comment]
The above solution still needs to check `modify_count`, otherwise, it may lead to over-estimation. Another proposal to maintain correct `modify_count`:
- each table and each index hold its own `modify_count`, and the value is modified in an incremental way as mentioned above;
- only allow one running analyze job for a specific object(i.e, a table, or an index) at a time, to prevent redundant `modify_count` subtraction by concurrent analyze jobs. Note that, concurrent analyze on a same object is just a waste of resources.

**Comment by winoros** [comment]
This is resolved when you set `@@tidb_analyze_version=2`.
`@@tidb_analyze_version=2` will be released soon as a final stable experiments.

**Comment by winoros** [comment]
fixed in release 5.2 and above
